tools/mmp/driver.cs

```diff
@@ -1340,6 +1339,10 @@ static int Compile ()
 
 					args.Append (StringUtils.Quote (lib)).Append (' ');
 
+					var libsystem_native_path = Path.Combine (libdir, "libmono-system-native.a");
+					args.Append (StringUtils.Quote (libsystem_native_path)).Append (' ');
+					args.Append ("-u ").Append ("_SystemNative_RealPath").Append (' '); // This keeps libmono_system_native_la-pal_io.o symbols
+
```

Comment:
I'm likely missing some context but, if this `.a` is always required then why is it not part of libmono itself ?

Even if it needs to be separate (that's worthy of a comment why) then what depends on it ? e.g. can we remove that extra step (size and build time) if the linker remove the code that requires it ?

Reply:
After talking with @alexischr this is, so far, only used for `FileSystemWatcher` with XM (the version in XI throws `NIE`).

For that specific case the above seems fine when we embed mono (it's static like libmono)

@chamons @rolfbjarne are there cases where `libmono-2.0.dylib` is used **without** the system mono ? (we do build and ship it)

Reply:
> After talking with @alexischr this is, so far, only used for FileSystemWatcher with XM (the version in XI throws NIE).

This doesn't look right according to which test failed:

    MonoTests.System.Diagnostics.EventLogTest.MachineName: System.InvalidOperationException : object_op
    ----> System.EntryPointNotFoundException : SystemNative_RealPath

Reply:
@spouliot no, when we embed mono in XM apps we always use the static library: https://github.com/xamarin/xamarin-macios/blob/10d98e67d88231fd41218a93b92d42914550ee10/tools/mmp/driver.cs#L1334-L1335

Reply:
We have [code](https://github.com/xamarin/xamarin-macios/blob/mono-2018-06/tools/linker/MonoTouch.Tuner/ListExportedSymbols.cs) to detect P/Invokes to `__Internal` functions and automatically add `-u _XXX` flags to the linker for such functions.

However,

1. IIRC this code only runs when embedding mono (because when using the system mono, we end up finding a bunch of non-existent symbols that are never used at runtime, and thus the native linker fails). Special cases should be added in the Target.CollectAllSymbols function: https://github.com/xamarin/xamarin-macios/blob/b1831013503655170455b2bcdd6b334b3e546f01/tools/common/Target.cs#L235
2. It also depends on the P/Invoke being declared as `__Internal`, which is not the case here. However: if we're using a static library, declaring the P/Invoke as `__Internal` is the right thing to do, so we should either fix the P/Invoke, or add a corresponding redirection in a dllmap and add support to our P/Invoke detection code to handle dllmaps.

Reply:
@rolfbjarne EventLog uses FileSystemWatcher, which is why the failure shows there. Nothing else in Mono uses the `SystemNative_` functions yet.

Regarding your other points:

1) I tried to do the special casing in  Target.CollectAllSymbols and it did not work, as I could not find how to only enable this when Mono is embedded (`embed_mono` in `Driver`). Otherwise, clang fails since the symbol will not be there when mmp is being used in the "dynamic mono" mode.

2) `System.Native` is used everywhere in CoreFX for pointing to the PAL library, it is not feasible to add #ifs everywhere to make it "_Internal" in Mono. When the P/Invoke detection code is ready to follow the dllmap to "_Internal", we should just use that, as this approach is not great long-term.

But this is a 2018-06 integration issue, so I am trying to unblock the integration and not give another task to the team :)

Reply:
@alexischr note that @rolfbjarne already approved the PR - but the discussion (more a reply to me) still needs to happen because this will break in the future (and block updating mono). Also I have **not** approved the PR because it must be green and that change requires updated the tests :)

Reply:
@alexischr: Only one #if should be needed: https://gist.github.com/rolfbjarne/8d1173f20f9baa58a4d4f2db596a9a45. That said, we should still implement support for dllmaps at some point.